File: Protobuf/src/101/files/cpp/reader.cpp
#include <iostream>
#include <fstream>
#include "messages.pb.h"

void PrintAddressBook(const tutorial::AddressBook& address_book) {
    for (int i = 0; i < address_book.people_size(); i++) {
        const tutorial::Person& person = address_book.people(i);
        
        std::cout << "Person ID: " << person.id() << std::endl;
        std::cout << "  Name: " << person.name() << std::endl;
        std::cout << "  E-mail: " << person.email() << std::endl;
        
        for (int j = 0; j < person.phones_size(); j++) {
            const tutorial::Person::PhoneNumber& phone = person.phones(j);
            
            switch (phone.type()) {
                case tutorial::Person::MOBILE:
                    std::cout << "  Mobile phone: ";
                    break;
                case tutorial::Person::HOME:
                    std::cout << "  Home phone: ";
                    break;
                case tutorial::Person::WORK:
                    std::cout << "  Work phone: ";
                    break;
            }
            std::cout << phone.number() << std::endl;
        }
    }
}

int main() {
    GOOGLE_PROTOBUF_VERIFY_VERSION;
    
    tutorial::AddressBook address_book;
    
    std::fstream input("addressbook.pb", std::ios::in | std::ios::binary);
    if (!input) {
        std::cerr << "File not found." << std::endl;
        return -1;
    } else if (!address_book.ParseFromIstream(&input)) {
        std::cerr << "Failed to parse address book." << std::endl;
        return -1;
    }
    
    PrintAddressBook(address_book);
    
    google::protobuf::ShutdownProtobufLibrary();
    
    return 0;
}

File: Protobuf/src/101/files/rust/reader.rs
use std::fs;
use prost::Message;

pub mod tutorial {
    include!(concat!(env!("OUT_DIR"), "/tutorial.rs"));
}

use tutorial::{AddressBook, person::PhoneType};

fn main() -> std::io::Result<()> {
    let data = fs::read("addressbook.pb")?;
    
    let address_book = AddressBook::decode(&data[..])
        .expect("Failed to parse address book");
    
    for person in address_book.people {
        println!("Person ID: {}", person.id);
        println!("  Name: {}", person.name);
        println!("  E-mail: {}", person.email);
        
        for phone in person.phones {
            let phone_type = match PhoneType::try_from(phone.r#type) {
                Ok(PhoneType::Mobile) => "Mobile phone",
                Ok(PhoneType::Home) => "Home phone",
                Ok(PhoneType::Work) => "Work phone",
                _ => "Unknown",
            };
            println!("  {}: {}", phone_type, phone.number);
        }
    }
    
    Ok(())
}
